detection_rules/integrations.py

```diff
@@ -0,0 +1,128 @@
+        if os.path.exists(manifest_file_path):
+            os.remove(manifest_file_path)
+    rule_integrations = os.listdir(INTEGRATION_RULE_DIR)
+    if "endpoint" in rule_integrations:
+        rule_integrations.remove("endpoint")
+
+    final_integration_manifests = {key: dict() for key in rule_integrations}
+
+    # initialize github client and point to package-storage prod
+    github = GithubClient(token)
+    client = github.authenticated_client
+    organization = client.get_organization("elastic")
+    repository = organization.get_repo("package-storage")
+    pkg_storage_prod_branch = repository.get_branch("production")
+    pkg_storage_branch_sha = pkg_storage_prod_branch.commit.sha
+
+    for ints in rule_integrations:
+        integration_manifests = get_integration_manifests(repository, pkg_storage_branch_sha,
+                                                          pkg_path=f"packages/{ints}")
+        for int_man in integration_manifests:
+            validated_manifest = IntegrationManifestSchema(unknown=EXCLUDE).load(int_man)
+            package_version = validated_manifest.pop("version")
+            final_integration_manifests[ints][package_version] = validated_manifest
+
+    manifest_file = gzip.open(manifest_file_path, "w+")
+    manifest_file_bytes = json.dumps(final_integration_manifests).encode("utf-8")
+    manifest_file.write(manifest_file_bytes)
+
+
+def find_least_compatible_version(package: str, integration: str,
+                                  current_stack_version: str, packages_manifest: dict) -> str:
+    """Finds least compatible version for specified integration based on stack version supplied."""
+    integration_manifests = packages_manifest[package]
+    least_compatible_version = None
+    compatible_versions = {}
+
+    def compare_versions(int_ver: str, pkg_ver: str) -> bool:
+        """Compares integration and package version"""
+        pkg_major, pkg_minor = Version(pkg_ver)
+        int_major, int_minor = Version(int_ver)[:2]
+
+        if int(int_major) < int(pkg_major) or int(pkg_major) > int(int_major):
+            return False
+        compatible = Version(int_ver) <= Version(pkg_ver)
+        return compatible
+
+    for ver, manifest in integration_manifests.items():
+        kibana_compat_vers = re.sub(r"\>|\<|\=|\^", "", manifest["conditions"]["kibana.version"])
+        kibana_compat_vers = kibana_compat_vers.split(" || ")
+        bool_checks = [compare_versions(kcv, current_stack_version) for kcv in kibana_compat_vers]
+        if any(bool_checks):
+            compatible_versions.setdefault(ver, True)
+
+    if len(list(compatible_versions.keys())) > 0:
+        compatible_version_list = list(compatible_versions.keys())
+        compatible_version_list.sort(key=Version)
+        least_compatible_version = compatible_version_list[0]
+    else:
+        raise Exception(f"no compatible version for integration {package}:{integration}")
+    return least_compatible_version
```

Comment:
I definitely see where you were going with this. Had to make a small adjustment though because `version` is a string and therefore cannot be sorted. We originally sorted the compatible version list to get the least compatible. In order to maintain this, I made the integration_manifests load as so:

```
integration_manifests = {k:v for k,v in sorted(packages_manifest[package].items(), key=Version)}
``

This will not only load the package but sort it as well by version so that when our loop execution occurs, the versions are already ordered in ascending value and we can return the first version found to be compatible.